jobs/periodic/delete-unreachable-achievements/courses.ts

```diff
@@ -0,0 +1,74 @@
+import { getLogger } from '../../../common/logger/logger';
+import { EventRelationType, createRelation, parseRelation } from '../../../common/achievement/relation';
+import { prisma } from '../../../common/prisma';
+import tracer from '../../../common/logger/tracing';
+import { subcourseOver } from '../../../common/courses/states';
+import { getSubcourse } from '../../../graphql/util';
+
+const logger = getLogger('CourseAchievementCleanup');
+
+const ParticipationGroupTemplate = 'pupil_course_participation';
+
+export const deleteUnreachableCourseAchievements = tracer.wrap('delete', _deleteUnreachableCourseAchievements);
+async function _deleteUnreachableCourseAchievements(dryRun: boolean = true) {
+    const till = new Date();
+    till.setDate(till.getDate() - 1);
+
+    const unachievedAchievementes = await prisma.user_achievement.findMany({
+        where: {
+            template: {
+                group: ParticipationGroupTemplate,
+            },
+            achievedAt: null,
+        },
+    });
+
+    const relationToUserIds: { [key: string]: string[] } = unachievedAchievementes.reduce((acc, curr) => {
+        if (!acc[curr.relation]) {
+            acc[curr.relation] = [];
+        }
+        acc[curr.relation].push(curr.userId);
+        return acc;
+    }, {});
+
+    logger.info('found possibly unachievable achievements', {
+        courses: relationToUserIds,
+        search_till: till,
+    });
+
+    for (const [relation, userIds] of Object.entries(relationToUserIds)) {
+        const [relationType, subcourseId] = parseRelation(relation);
+        if (relationType !== EventRelationType.Subcourse) {
+            logger.warn('unexpected relation type', {
+                relation,
+            });
+            continue;
+        }
+
+        const subcourse = await getSubcourse(subcourseId, false);
+        if (subcourse.cancelled || !(await subcourseOver(subcourse))) {
```

Comment:
Unfortunately, `subcourseOver` needs the whole subcourse even tho it only uses the subcourse.id.

I thought about refactoring it a little, but I don't know about potential side effects.